### PR TITLE
add vim/helix `~`

### PR DESCRIPTION
**Pull request by xdBronch**
this doesnt match vim perfectly since there it moves you to the beginning of the selection in visual mode which i couldnt quite figure out, though i didnt try super hard. otherwise i believe this matches both

### CODE REVIEW

File: src/tui/editor.zig [hunk 1]
```diff
@@ -3924,6 +3925,49 @@ pub const Editor = struct {
...
+            defer bytes.deinit();
+
+            var iter = code_point.Iterator{ .bytes = cut_text };
+            var buf: [4]u8 = undefined;
+
+            while (iter.next()) |cp| {
+                const code = if (cd.isLower(cp.code)) cd.toUpper(cp.code) else cd.toLower(cp.code);
+                const len = std.unicode.utf8Encode(code, &buf) catch return error.Stop;
+                bytes.appendSliceAssumeCapacity(buf[0..len]);
+            }
```
Comment (neurocyte) [inline src/tui/editor.zig:3952]: You can do codepoint iteration directly in the Buffer with root.walk_egc_forward without first copying the selection, and without adding an additional module. It's not a big deal, but the way it's written here seems a little over complex and it walks twice, once in copy_selection and again with code_point.Iterator.

File: src/tui/mode/input/vim/visual.zig [hunk 2]
```diff
@@ -208,6 +208,7 @@ fn mapPress(self: *Self, keypress: u32, egc: u32, modifiers: u32) !void {
 
             'o' => self.seq(.{ "smart_insert_line_before", "enter_mode" }, command.fmt(.{"vim/insert"})),
 
+            '`' => self.cmd("switch_case", .{}),
```
Comment (neurocyte) [inline src/tui/mode/input/vim/visual.zig:211]: Why would this not be identical to normal mode?

File: src/tui/editor.zig [hunk 3]
```diff
@@ -3924,6 +3925,49 @@ pub const Editor = struct {
...
+            var bytes = std.ArrayList(u8).initCapacity(allocator, cut_text.len) catch return error.Stop;
+            defer bytes.deinit();
+
+            var iter = code_point.Iterator{ .bytes = cut_text };
+            var buf: [4]u8 = undefined;
+
+            while (iter.next()) |cp| {
+                const code = if (cd.isLower(cp.code)) cd.toUpper(cp.code) else cd.toLower(cp.code);
+                const len = std.unicode.utf8Encode(code, &buf) catch return error.Stop;
+                bytes.appendSliceAssumeCapacity(buf[0..len]);
```
Comment (neurocyte) [inline src/tui/editor.zig:3951]: This seems likely to panic. Who guarantees that the converted codepoints encode to exactly the same total byte length as the original string?

**Review by neurocyte — CHANGES_REQUESTED** [review]
I ran `zig fmt` on all source files, so you can remove that commit here.

File: src/tui/editor.zig [hunk 4]
```diff
@@ -3924,6 +3925,49 @@ pub const Editor = struct {
     }
     pub const to_lower_meta = .{ .description = "Convert selection or word to lower case" };
 
+    fn switch_case_cursel(self: *Self, root_: Buffer.Root, cursel: *CurSel, allocator: Allocator) error{Stop}!Buffer.Root {
+        var root = root_;
+        const saved = cursel.*;
+        const sel = if (cursel.selection) |*sel| sel else ret: {
+            var sel = cursel.enable_selection();
```
Comment (neurocyte) [inline src/tui/editor.zig:3932]: Should the selection perhaps be disabled again if it was added automatically here?

File: src/tui/mode/input/vim/visual.zig (hunk 2 above)
Comment (xdBronch) [reply src/tui/mode/input/vim/visual.zig:211]: thats how it works in vim, in visual mode it just switches your selection and moves your cursor back to the beginning, no advancing the cursor
<https://asciinema.org/a/aIgiFI9X6ojghM0J1nzubjP4J>

Comment (neurocyte) [reply src/tui/mode/input/vim/visual.zig:211]: ok

File: src/tui/editor.zig (hunk 1 above)
Comment (xdBronch) [reply src/tui/editor.zig:3952]: sorry, im having a bit of trouble getting this to work :/. how important is it to change how this is done? `to_lower` and `to_upper` are doing nearly the same thing except `zg` has a function that does it

Comment (neurocyte) [reply src/tui/editor.zig:3952]: I can rewrite it to use walk_egc for you if you want.

Comment (xdBronch) [reply src/tui/editor.zig:3952]: if you dont mind that would be great, thank you

**Comment by neurocyte** [comment]
I decided to keep it simple and used the same pattern as to_upper/lower.

**Pull request merged**